src/zenml/integrations/vllm/services/vllm_deployment.py

```diff
@@ -0,0 +1,142 @@
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at:
+#
+#       https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+#  or implied. See the License for the specific language governing
+#  permissions and limitations under the License.
+"""Implementation of the vLLM Inference Server Service."""
+
+from typing import Any, List, Optional, Union
+
+from zenml.logger import get_logger
+from zenml.services import (
+    LocalDaemonService,
+    LocalDaemonServiceConfig,
+    ServiceType,
+)
+from zenml.services.service import BaseDeploymentService
+
+logger = get_logger(__name__)
+
+
+class VLLMServiceConfig(LocalDaemonServiceConfig):
+    """vLLM service configurations."""
+
+    blocking: bool = True
+    model: Optional[str] = None
+    # If unspecified, model name or path will be used.
+    tokenizer: Optional[str] = None
+    served_model_name: Optional[Union[str, List[str]]] = None
+    # Trust remote code from huggingface.
+    trust_remote_code: Optional[bool] = False
+    # ['auto', 'slow', 'mistral']
+    tokenizer_mode: Optional[str] = "auto"
+    # ['auto', 'half', 'float16', 'bfloat16', 'float', 'float32']
+    dtype: Optional[str] = "auto"
+    # The specific model version to use. It can be a branch name, a tag name, or a commit id.
+    # If unspecified, will use the default version.
+    revision: Optional[str] = None
+
+
+class VLLMDeploymentService(LocalDaemonService, BaseDeploymentService):
+    """vLLM Inference Server Deployment Service."""
+
+    SERVICE_TYPE = ServiceType(
+        name="vllm-deployment",
+        type="model-serving",
+        flavor="vllm",
+        description="vLLM Inference prediction service",
+    )
+    config: VLLMServiceConfig
+
+    def __init__(self, config: VLLMServiceConfig, **attrs: Any):
+        """Initialize the vLLM deployment service.
+
+        Args:
+            config: service configuration
+            attrs: additional attributes to set on the service
+        """
+        super().__init__(config=config, **attrs)
+
+    def run(self) -> None:
+        """Start the service."""
+        logger.info(
+            "Starting vLLM inference server service as blocking "
+            "process... press CTRL+C once to stop it."
+        )
+
+        import uvloop
+        from vllm.entrypoints.openai.api_server import run_server
+        from vllm.entrypoints.openai.cli_args import make_arg_parser
+        from vllm.utils import FlexibleArgumentParser
+
+        try:
+            parser = make_arg_parser(FlexibleArgumentParser())
+            args = parser.parse_args()
+            # Update the arguments in place
+            args.__dict__.update(self.config.model_dump())
+            uvloop.run(run_server(args=args))
```

Comment:
One of the main added things in the `LocalDaemonServiceEndpoint` is looking for a free port in the local environment which then can be used when starting the server to avoid problems when the main given port is locked to some other process. [check this](https://github.com/zenml-io/zenml/blob/main/src/zenml/services/local/local_service_endpoint.py#L88).

we can then add port to the args to make sure that our server is starting on it, example [here](https://github.com/zenml-io/zenml/blob/main/src/zenml/integrations/mlflow/services/mlflow_deployment.py#L245)

Reply:
Thanks @safoinme  for the pointers.

I have updated the code to use endpoint configuration.

Reply:
Also updated the example on `zenml-projects` repo and tested the example.